From 6ccebf2aacb0b2bcb3e4729b083fd81ce4ff53df Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Sat, 2 Mar 2024 21:19:25 +0100
Subject: [PATCH 38/46] [HACK] mfmediaengine: Do not send
 MF_MEDIA_ENGINE_EVENT_ERROR for VRChat.

Breaks AVPro Video (used by VRChat) when a IMFMediaEngine::GetError call inside the event handler doesn't return an
IMFMediaError instance. The event handler code then tries to call a function on the non-existant IMFMediaError instance,
which segfaults.

I don't know why IMFMediaEngine::GetError returns a NULL instance inside the error callback. This should be fixed
properly. In the meantime, this hack works around the problem.
---
 dlls/mfmediaengine/main.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/dlls/mfmediaengine/main.c b/dlls/mfmediaengine/main.c
index fba338a8438..2986c75902f 100644
--- a/dlls/mfmediaengine/main.c
+++ b/dlls/mfmediaengine/main.c
@@ -1466,11 +1466,21 @@ static HRESULT WINAPI media_engine_load_handler_Invoke(IMFAsyncCallback *iface,
     }
     else
     {
-        engine->network_state = MF_MEDIA_ENGINE_NETWORK_NO_SOURCE;
-        engine->error_code = MF_MEDIA_ENGINE_ERR_SRC_NOT_SUPPORTED;
-        engine->extended_code = hr;
-        IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_ERROR, engine->error_code,
-                engine->extended_code);
+        static unsigned int once;
+        char sgi[64];
+
+        if (GetEnvironmentVariableA("SteamGameId", sgi, sizeof(sgi)) && !strcmp(sgi, "438100"))
+        {
+            if (!once++) FIXME("HACK: not sending MF_MEDIA_ENGINE_EVENT_ERROR event.\n");
+        }
+        else
+        {
+            engine->network_state = MF_MEDIA_ENGINE_NETWORK_NO_SOURCE;
+            engine->error_code = MF_MEDIA_ENGINE_ERR_SRC_NOT_SUPPORTED;
+            engine->extended_code = hr;
+            IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_ERROR, engine->error_code,
+                    engine->extended_code);
+        }
     }
 
     LeaveCriticalSection(&engine->cs);
-- 
2.46.0

